src/Codegen/Typing/TypeSystem.hack

```diff
@@ -182,10 +180,90 @@ final abstract class TypeSystem {
     // use the alias A instead of using the solved type T. This better matches developer
     // intent when combining aliases and other types: the union of A and a type T2
     // which is itself a subtype of A should be expressed as A.
-    if (C\count(Vec\unique_by($equivalent_types, $ty ==> $ty->render())) === 1) {
+    if (C\count(Vec\unique_by($equivalent_types, $ty ==> $ty->getName())) === 1) {
       return C\firstx($equivalent_types);
     } else {
       return $type;
     }
   }
-}
+
+  private static function resolveGenerics(ConcreteTypeName $builtin_type_name, vec<Type> $types): vec<Type> {
+    $generics = vec[];
+    // TODO: Handle dicts
+    if ($builtin_type_name === ConcreteTypeName::VEC || $builtin_type_name === ConcreteTypeName::KEYSET) {
+      $generics[] = Vec\map($types, $ty ==> C\first($ty->getGenerics()))
+        |> Vec\filter_nulls($$)
+        |> self::union($$);
+    }
+    return $generics;
+  }
+
+  private static function resolveShapeFields(ConcreteTypeName $builtin_type_name, vec<Type> $types): (Type::TShapeFields, bool) {
+    // Union of shapes: union the shape fields such that `shape('a' => int) | shape('a' => string)`
+    // becomes `shape('a' => arraykey)`. Assume any field absent from an open shape can only be
+    // typed as `mixed`, so that `shape('a' => int) | shape('b' => int, ...)` becomes
+    // `shape(?'b' => int, ...)`.
+    $is_closed_shape = false;
+    $shape_fields = dict[];
+
+    if ($builtin_type_name === ConcreteTypeName::SHAPE) {
+      // All types must be shapes or subtypes of shapes (i.e. nothing). In generating our unioned shape,
+      // we only consider shapes, skipping nothing types.
+      $num_shapes = 0;
+      $num_closed_shapes = 0;
+      $shape_field_types = dict[];
+      $open_shape_field_counts = dict[];
+      $required_field_counts = dict[];
+
+      foreach ($types as $ty) {
+        if ($ty->getConcreteTypeName() === ConcreteTypeName::SHAPE) {
+          $num_shapes++;
+          if ($ty->isClosedShape()) {
+            $num_closed_shapes++;
+          }
+
+          foreach ($ty->getShapeFields() as $name => $field) {
```

Comment:
how does this handle shapes within shapes? these just get listed out as "shape" types?

Reply:
Yup, exactly. If multiple shapes contain a field with the same name, we'll union the types of that field recursively, here: https://github.com/slackhq/hack-json-schema/pull/78/files#diff-18da79e1d9a43454019fd566249550acdfcc9397cd8f5b915ca806ff7f053a44R248

That's how we handle nested shapes as well.